realm/realm-library/src/main/java/io/realm/sync/Subscription.java

```diff
@@ -0,0 +1,228 @@
+         */
+        ERROR((byte) -1),
+
+        /**
+         * The subscription has been created, but has not yet been processed by the sync
+         * server.
+         */
+        PENDING((byte) 0),
+
+        /**
+         * The subscription has been processed by the Realm Object Server and data is being synced
+         * to the device.
+         */
+        ACTIVE((byte) 1),
+
+        /**
+         * The subscription has been removed. Data is no longer being synchronized from the Realm
+         * Object Server, and the objects covered by this subscription might be deleted from the
+         * device if no other subscriptions include them.
+         */
+        INVALIDATED(null);
+
+
+        private final Byte nativeValue;
+
+        State(Byte nativeValue) {
+            this.nativeValue = nativeValue;
+        }
+
+        /**
+         * Returns the native value representing this state.
+         *
+         * @return the native value representing this state.
+         */
+        public Byte getValue() {
+            return nativeValue;
+        }
+    }
+
+    public Subscription() {
+        // Required by Realm.
+    }
+
+    /**
+     * Creates a unmanaged named subscription from a {@link RealmQuery}.
+     * This will not take effect until it has been added to the Realm.
+     *
+     * @param name name of the query.
+     * @param query the query to turn into a subscription.
+     */
+    public Subscription(String name, RealmQuery<?> query) {
+        this.name = name;
+        this.query = query.getDescription();
+        this.status = 0;
+        this.errorMessage = "";
+        this.matchesProperty = "";
+    }
+
+    @Index
+    @Required
+    private String name;
+
+    /**
+     * The underlying representation of the State
+     */
+    private byte status;
+
+    @Required
+    @RealmField("error_message")
+    private String errorMessage;
+
+    @Required
+    @RealmField("matches_property")
+    private String matchesProperty;
+
+    @Required
+    private String query;
+
+    @RealmField("query_parse_counter")
+    private int queryParseCounter;
+
+    /**
+     * Returns the name of the subscription.
+     *
+     * @return the name of the subscription.
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Returns a textual description of the query that created this subscription.
+     *
+     * @return a textual description of the query
+     */
+    public String getQueryDescription() {
+        return query;
+    }
+
+    /**
+     * Returns the internal name of the Class being queried.
+     *
+     * @return the internal name of the of the class being queried.
+     */
+    public String getQueryClassName() {
+        // Strip the __matches suffix to end up with the class being queried.
+        String classQueried = matchesProperty;
+        return classQueried.substring(0, classQueried.length() - "_matches".length());
```

Comment:
This might not always give the expected result. Please take a look at https://github.com/realm/realm-js/pull/2075

Reply:
Actually, this should always be safe as we created the `*_matches` property so even if the user had a class named `foo_matches` it would become `foo_matches_matches` here and the above way only remove the last `_matches`